File: cpp_lab4/MyStack.cpp
//
// Created by chise on 27-Sep-18.
//

#include <iostream>
#include "MyStack.h"

int MyStack::size() {
    return stack.size();
}

void MyStack::push(char c) {
    stack = stack + c;
}

char MyStack::pop() {
    if (stack.size() == 0)
        return 0;
    char c = stack[stack.size() - 1];
    stack = stack.substr(0, stack.size() - 1);
    return c;
}

bool MyStack::isFull() {
    return size() == maxSize;
}

bool MyStack::isEmpty() {
    return size() == 0;
}

MyStack::MyStack(std::string s) {

}

MyStack::MyStack() {

}

void MyStack::print() {
    std::cout << stack.c_str() << std::endl;
}

File: laburi_java/src/lab4/task1/MyStack.java
package lab4.task1;

import java.util.Arrays;
import java.util.Objects;

public class MyStack {
    private int size;
    private int maxSize;
    private char list[];

    public MyStack() {
        this(10);
    }

    public MyStack(int maxSize) {
        this.maxSize = maxSize;
        this.size = 0;
    }

    public boolean push(char value) {
        if (isFull()) {
            return false;
        }
        reallocList(size + 1);
        list[size - 1] = value;
        return true;
    }

    public boolean isFull() {
        return size == maxSize;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public char pop() {
        if (isEmpty()) {
            return 0;
        }
        char value = list[size - 1];
        reallocList(size - 1);
        return value;
    }

    private void reallocList(int newSize) {
        char newList[] = new char[newSize];
        if (!Objects.isNull(list)) {
            int start = size > newSize ? 1 : 0;
            System.arraycopy(list, start, newList, 0, size > newSize ? newSize : size);
        }
        size = newSize;
        list = newList;
    }

    public void showList() {
        if (!Objects.isNull(list)) {
            System.out.println(Arrays.toString(list));
        } else {
            System.out.println("List is null");
        }
    }
}
